host/fs/fs_linux.go

```diff
@@ -114,3 +118,259 @@ func (e *event) wait(timeoutms int) (int, error) {
+			// maybe not worth it.
+			if ep&epollERR != 0 {
+				// TODO(maruel): We should stop listening to this file descriptor.
+				// Same for epollHUP.
+			}
+			// pipe and sockets trigger epollIN and epollOUT, but GPIO sysfs
+			// triggers epollPRI.
+			if ep&(epollPRI|epollIN|epollOUT) != 0 {
+				lookups = append(lookups, lookup{c: c, event: ep})
+			}
+		}
+		e.mu.Unlock()
+
+		// Once the lock is released, send the events. It's up to the client to
+		// interpret the epoll flag.
+		for _, t := range lookups {
+			t.c <- now
+		}
+	}
+}
+
+// addFd starts listening to events generated by file descriptor |fd|.
+//
+// fd is the OS file descriptor. In practice, it must fit a int32 value. It
+// works on pipes, sockets and sysfs objects like GPIO but not on standard
+// files.
+//
+// c is the channel to send events to. Unbuffered channel will block the event
+// loop, which may mean lost events, especially if multiple files are listened
+// to simultaneously.
+//
+// flags is the events to listen to. No need to specify epollERR and epollHUP,
+// they are sent anyway.
+//
+// addFd lazy initializes eventsListener if it was not initialized yet.
+//
+// It can fail due to various reasons, a few are:
+//   ENOSPC: /proc/sys/fs/epoll/max_user_watches limit was exceeded
+//   ENOMEM: No memory available
+//   EPERM: fd is a regular file or directory
+func (e *eventsListener) addFd(fd uintptr, c chan<- time.Time, flags epollEvent) error {
+	if c == nil {
+		return errors.New("fd: addFd requires a valid channel")
+	}
+	if err := e.init(); err != nil {
+		return err
+	}
+	if err := e.addFdInner(fd, flags); err != nil {
+		return err
+	}
+	e.mu.Lock()
+	e.fds[int32(fd)] = c
+	e.mu.Unlock()
+	// Wake up the poller so it notices there's one new file.
+	e.wakeUpLoop(nil)
+	return nil
+}
+
+func (e *eventsListener) addFdInner(fd uintptr, flags epollEvent) error {
+	ev := syscall.EpollEvent{Events: uint32(flags), Fd: int32(fd)}
+	return syscall.EpollCtl(e.epollFd, epollCTLAdd, int(fd), &ev)
+}
+
+// removeFd stop listening to events on this file descriptor.
+func (e *eventsListener) removeFd(fd uintptr) error {
+	if err := syscall.EpollCtl(e.epollFd, epollCTLDel, int(fd), nil); err != nil {
+		return err
+	}
+	e.mu.Lock()
+	delete(e.fds, int32(fd))
+	e.mu.Unlock()
+	// Wake up the poller so it notices there's one less file.
+	e.wakeUpLoop(nil)
+	return nil
+}
+
+// wakeUpLoop wakes up the poller and waits for it.
+//
+// Must not be called with the lock held.
+func (e *eventsListener) wakeUpLoop(c <-chan time.Time) time.Time {
+	e.mu.Lock()
+	running := e.epollFd != 0
+	e.mu.Unlock()
+	if !running {
+		// The loop is not running, nothing to wake up.
+		return time.Now()
+	}
+	// TODO(maruel): Figure out a way to wake up that doesn't require emptying.
+	var b [1]byte
+	_, _ = e.w.Write(b[:])
+	var t time.Time
+	if c != nil {
+		// To prevent deadlock, also empty c.
+		for {
+			select {
+			case <-c:
+			case t = <-e.wakeUp:
```

Comment:
Then this would be failure. :/

Reply:
I mean this would be because of a logic failure or that the loop terminated unexpectedly. The problem right now is that the loop doesn't send the events in a goroutine, and the the channels can be blocking.

Reply:
Its minor as only way I could see a possible hang in normal operation was with a Close() at exactly the wrong time. Since it not supposed to used outside tests I don't think its a bit deal, its just maybe little fragile.

Reply:
I renamed Close to stopLoop.

I found a better fix; stopLoop now errors if there is an active listener.